lib/recursive-methods.rb

```diff
@@ -1,49 +1,140 @@
 # Authoring recursive algorithms. Add comments including time and space complexity for each method.
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity:  O(n)  
+# Space complexity: O(1)
+
 def factorial(n)
-    raise NotImplementedError, "Method not implemented"
+    # if wrong argument is provided
+    if n < 0
+        raise ArgumentError
+    end
+    #base case
+    if n == 0
+        return 1
+    else
+        return n * factorial(n-1)
+    end
 end
 
-# Time complexity: ?
-# Space complexity: ?
+
+# Time complexity: O(n)
+# Space complexity: O(n)
 def reverse(s)
```

Comment:
Since you are making a new string with each recursive call this is O(n^2) space complexity.